File: 164.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int maximumGap(vector<int>& nums) {
        sort(nums, 0, nums.size() - 1);
        int max = 0;
        for (int i = 1; i < nums.size(); ++i) {
            max = max < nums[i] - nums[i-1] ? nums[i] - nums[i-1] : max;
        }
        return max;
    }
    int divide(vector<int> &nums, int low, int high) {
        int pivot = nums[low];
        while (low < high) {
            while (high > low && nums[high] >= pivot) --high;
            if (high > low) {
                int temp = nums[low];
                nums[low] = nums[high];
                nums[high] = temp;
            }
            while (high > low && nums[low] <= pivot) ++low;
            if (high > low) {
                int temp = nums[low];
                nums[low] = nums[high];
                nums[high] = temp;
            }
        }
        nums[low] = pivot;
        return low;
    }
    void sort(vector<int> &nums, int low, int high) {
        if (low >= high) return;
        int mid = divide(nums, low, high);
        sort(nums, low, mid - 1);
        sort(nums, mid + 1, high);
    }
};

File: 164.go
package leetcode

func maximumGap(nums []int) int {
    if (len(nums) < 2) {
        return 0
    }
	nums = quicksort(nums, 0, len(nums)-1)
	max := 0
	for i := 1; i < len(nums);  i++ {
		if (nums[i] - nums[i-1] > max) {
			max = nums[i] - nums[i-1]
		}
	}
	return max
}

func divide(nums []int, low int, high int) int {
	pivot := nums[low]
	for low < high {
		for low < high && nums[high] >= pivot {
			high--
		}
		if (low < high) {
			temp := nums[low]
			nums[low] = nums[high]
			nums[high] = temp
		}
		for low < high && nums[low] <= pivot {
			low++
		}
        if (low < high) {
			temp := nums[low]
			nums[low] = nums[high]
			nums[high] = temp
		}
	}
	nums[low] = pivot
	return low
}

func quicksort(nums []int, low int, high int) []int{
	if (low >= high) {
		return nums
	}
	mid := divide(nums, low, high)
	quicksort(nums, low, mid - 1)
	quicksort(nums, mid + 1, high)
	return nums
}